### PR TITLE
Breaking: report fatal error for linting nonexistent files (fixes #7390)

### LINKED ISSUE #7390: eslint silently ignores missing files
- **ESLint Version:** 3.8.1
- **Node Version:** 6.3.1
- **npm Version:** 3.10.3

**Please show your full configuration:**

none (default)

**What did you do? Please include the actual source code causing the issue.**

``` sh
$ eslint missing-file
```

**What did you expect to happen?**

A nonzero exit status, and an error. It sounds like this used to happen:
- https://github.com/eslint/eslint/issues/1296
- https://github.com/eslint/eslint/issues/1831

**What actually happened? Please include the actual, raw output from ESLint.**

No output, 0 exit status.

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Bug fix (#7390)

**What changes did you make? (Give an overview)**

This is the same as https://github.com/eslint/eslint/pull/7395, but I rebased it onto master and modified the error message a bit. As discussed in #7390, this updates ESLint core to report a fatal error whenever any of the globs passed on the CLI do not match any files.

**Is there anything you'd like reviewers to focus on?**

Nothing in particular

### CODE REVIEW

**Comment by not-an-aardvark** [comment]
I noticed that there were some differences between the implementation from https://github.com/eslint/eslint/pull/7395 and the proposal from the TSC meeting:

* Providing an empty directory as a path would not result in an error.
* Providing a glob where all files are ignored would not result in an error.

I updated this PR to make both of the above cases errors.

File: tests/lib/cli-engine.js [hunk 1]
```diff
@@ -2566,6 +2566,43 @@ describe("CLIEngine", () => {
...
+            });
+
+            it("one file", () => {
+                assert.throws(() => {
+                    engine.executeOnFiles(["non-exist.js"]);
+                }, "'non-exist.js' was not found.");
+            });
+
+            it("should not throw if the directory exists", () => {
+                engine.executeOnFiles(["empty"]);
```
Comment (platinumazure) [inline tests/lib/cli-engine.js]: Could we `assert.doesNotThrow` just to be explicit about what we're asserting on? Thanks!

File: tests/lib/util/glob-util.js [hunk 2]
```diff
@@ -233,11 +223,12 @@ describe("globUtil", () => {
         it("should silently ignore default ignored files if not passed explicitly", () => {
```
Comment (platinumazure) [inline tests/lib/util/glob-util.js]: Is this test name still accurate given that we have an `assert.throws` in the test body?

File: tests/lib/util/glob-util.js [hunk 3]
```diff
@@ -254,12 +245,13 @@ describe("globUtil", () => {
         it("should silently ignore default ignored files if not passed explicitly even if ignore is false", () => {
```
Comment (platinumazure) [inline tests/lib/util/glob-util.js]: Is this test name still accurate given that we have an `assert.throws` in the test body?

File: tests/lib/util/glob-util.js [hunk 4]
```diff
@@ -275,19 +267,45 @@ describe("globUtil", () => {
...
+        it("should throw if only ignored files match a glob", () => {
+            const pattern = getFixturePath("glob-util", "ignored");
+            const options = { ignore: true, ignorePath: getFixturePath("glob-util", "ignored", ".eslintignore") };
+
+            assert.throws(() => {
+                globUtil.listFilesToProcess([pattern], options);
+            }, `All files matched by '${pattern}' are ignored.`);
         });
 
         it("should not return an ignored file", () => {
```
Comment (platinumazure) [inline tests/lib/util/glob-util.js]: Is this test name still accurate given that we have an `assert.throws` in the test body?

File: tests/lib/util/glob-util.js [hunk 5]
```diff
@@ -301,17 +319,19 @@ describe("globUtil", () => {
         it("should not return a file from a glob if it matches a pattern in an ignore file", () => {
```
Comment (platinumazure) [inline tests/lib/util/glob-util.js]: Is this test name still accurate given that we have an `assert.throws` in the test body?

File: tests/lib/util/glob-util.js [hunk 6]
```diff
@@ -301,17 +319,19 @@ describe("globUtil", () => {
...
             const patterns = [getFixturePath("glob-util", "ignored", "**/*.js")];
-            const result = globUtil.listFilesToProcess(patterns, options);
 
-            assert.strictEqual(result.length, 0);
+            assert.throws(() => {
+                globUtil.listFilesToProcess(patterns, options);
+            }, `All files matched by '${patterns[0]}' are ignored.`);
         });
 
         it("should not return a file from a glob if matching a specified ignore pattern", () => {
```
Comment (platinumazure) [inline tests/lib/util/glob-util.js]: Is this test name still accurate given that we have an `assert.throws` in the test body?

File: tests/lib/cli-engine.js (hunk 1 above)
Comment (platinumazure) [reply tests/lib/cli-engine.js]: I would, but I'll defer to consensus. I like `assert.doesNotThrow` for its explicitness if nothing else. But I realize I may be in the minority here.

Comment (not-an-aardvark) [reply tests/lib/cli-engine.js]: I suppose I'll add it now, and it can be removed later if we decide to do that.

Comment (not-an-aardvark) [reply tests/lib/cli-engine.js]: Actually, I just realized the question is moot because this should be throwing anyway. 😄

**Comment by not-an-aardvark** [comment]
Updated to fix all the failing tests and improve an error message.

**Pull request merged**